### DISCUSSION TITLE
Get Users for Tenant

### DISCUSSION THREAD

**Discussion by TheFrankman** [filamentphp/filament#10359] [Help] [answered]
I have setup mutli-tenancy in it's most basic form. The only difference is that i am using an Organisation::class rather than Team::class.

Everything works correctly, for my belongs to relationships i.e. let's say an artwork belongs to a given organisation via an organisation_id - Great.

However, i've come to the stage where i'd like to be able to create User Management. So i've created a UserResource for this.

When attempting to load the users table page i get the error : 

```
The model [App\Models\User] does not have a relationship named [organisation]. You can change the relationship being used by passing it to the [ownershipRelationship] argument of the [tenant()] method in configuration.
```

Which makes sense since a user belongs to many organisations and an organisation belongs to many users. So there is no 'organisation' relationship on my User model.

So, how are you meant to approach user management for multi-tenancy, or in this case, even just list the users that have access to the current tenant in a filament table ? 

Here is some further context : 

```
class AdminPanelProvider extends PanelProvider
{
    public function panel(Panel $panel): Panel
    {
        return $panel
            ->tenant(model:Organisation::class, slugAttribute: 'slug')
            ->tenantRegistration(page: RegisterOrganisation::class)
            ->tenantProfile(EditOrganisationProfile::class)
```

```
class User extends Authenticatable implements FilamentUser, HasTenants
{
    use HasApiTokens, HasFactory, Notifiable;
    
     public function organisations(): BelongsToMany
    {
        return $this->belongsToMany(Organisation::class);
    }

    public function getTenants(Panel $panel): array|Collection
    {
        return $this->organisations;
    }

    public function canAccessTenant(Model $tenant): bool
    {
        return $this->organisations->contains($tenant);
    }

    public function canAccessPanel(Panel $panel): bool
    {
        return true;
    }
```

```
class Organisation extends Model
{
    use HasFactory;

    public function users(): BelongsToMany
    {
        return $this->belongsToMany(User::class);
    }

    protected static function boot()
    {
        parent::boot();

        // Listen for the creating event and set the slug before saving
        static::creating(function (Organisation $organisation) {
            $organisation->slug = Str::slug($organisation->name);
        });
    }
}
```

```
class UserResource extends Resource
{
    protected static ?string $model = User::class;

    protected static ?string $navigationIcon = 'heroicon-o-rectangle-stack';

    public static function form(Form $form): Form
    {
        return $form
            ->schema([
                //
            ]);
    }

    public static function table(Table $table): Table
    {
        return $table
            ->modifyQueryUsing(function (Builder $query) {
                return $query->whereHas(Filament::getTenant()); // doesn't get this far.
            })
            ->columns([
                //
            ])
```

**Comment by TheFrankman** [comment]
I'm not going to mark this as the answer because it doesn't really feel right to me, but incase no-one helps me, and someone else stumbles upon this. I had some success by adding `protected static ?string $tenantOwnershipRelationshipName = 'organisations';` to my UserResource.

Not seeing any negative consequences so far.

**Comment by kombiahrk** [comment]
->tenant(Organisation::class,ownershipRelationship:'organisations',slugAttribute:'slug')
 
 Dont need to mention ownershiprelation in each resource

**Reply by TheFrankman** [reply]
Thanks fro the response @kombiahrk but i think you misunderstood.

The tenant relationship in general works with my modal relationships without `$tenantOwnershipRelationshipName`

It's just the users resource because users don't have an organisation_id and therefore don't have an organisation (singular) relationship. It's a many to many between organisations and users.

When using the UserResource and trying to render a table of users that's when i get the error i mentioned. Using `$tenantOwnershipRelationshipName` on the users resource allows me to say that the relationship is `organisations` plural, rather than `organisation` singular.

In my use case, i think this kind of makes sense because every other model has a belongsTo relationship with a specific organisation. It's only Users that can belong to many.

While this works, i feel like it must be a pretty common issue, but it isn't documented. Wondering if my approach is wrong.

**Reply by FabianWesner** [reply, marked as answer]
I have the same problem. My workaround is to disable the check on the UserResource:

```
protected static bool $isScopedToTenant = false;
``` 

and to filter manually:

```
    public static function getEloquentQuery(): Builder
    {
        $currentTeamId = Filament::getTenant()->id;

        return parent::getEloquentQuery()
                 ->whereHas('teams', function ($query) use ($currentTeamId) {
                $query->where('teams.id', $currentTeamId);
            });
    }
```

Does this work for you?

**Reply by FabianWesner** [reply]
but this works as well :-)

```
    protected static ?string $tenantOwnershipRelationshipName = 'teams';
```

**Reply by TheFrankman** [reply]
Thanks @FabianWesner glad to hear someone else is facing the same issue and not just me doing something silly !

**Reply by karlebh** [reply]
> I have the same problem. My workaround is to disable the check on the UserResource:
> 
> ```
> protected static bool $isScopedToTenant = false;
> ```
> 
> and to filter manually:
> 
> ```
>     public static function getEloquentQuery(): Builder
>     {
>         $currentTeamId = Filament::getTenant()->id;
> 
>         return parent::getEloquentQuery()
>                  ->whereHas('teams', function ($query) use ($currentTeamId) {
>                 $query->where('teams.id', $currentTeamId);
>             });
>     }
> ```
> 
> Does this work for you?

Life saver

**Comment by micraux** [comment]
Just for my information, and because I did this in my project...

Why didn't you create a `organisation_user` pivot table in you database?

**Reply by TheFrankman** [reply]
I do have an `organisation_user` - this was a little while ago now, but i believe this issue is that the tenancy functionality isn't expecting a that kind of relationship.

**Comment by dennyvik** [comment]
This is what is work for me:

in my case, i am using Branch model instead of Team, i set up a tenant-aware middleware to apply global scopes as mention in Filament Docs, and modified a little bit. Work with multi panel that not applied multi-tenancy, they listed all the users. while in the branch panel, the user list only the users that belongs to the same branch as the logged user's branch. 

I even using the same userResource that i used in other panel (non multi-tenancy panel). 

ApplyTenantScopes.php
```php
use App\Models\User;
use Closure;
use Filament\Facades\Filament;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Http\Request;
use Symfony\Component\HttpFoundation\Response;

class ApplyTenantScopes
{
    /**
     * Handle an incoming request.
     *
     * @param  \Closure(\Illuminate\Http\Request): (\Symfony\Component\HttpFoundation\Response)  $next
     */
    public function handle(Request $request, Closure $next): Response
    {
        User::addGlobalScope(
            fn (Builder $query) => $query->whereHas('branches', function ($query) {
                $query->where('branches.id', Filament::getTenant()->id);
                }
            )
        );
        
        return $next($request);
    }
}
```

Model User
```php
    public function branches(): BelongsToMany
    {
        return $this->belongsToMany(Branch::class)->withTimestamps();
    }
    
    public function getTenants(Panel $panel): Collection
    {
        return $this->branches;
    }
 
    public function canAccessTenant(Model $tenant): bool
    {
        return $this->branches()->whereKey($tenant)->exists();
    }
```

Model Branch
```php
    public function users():BelongsToMany
    {
        return $this->belongsToMany(User::class)->withTimestamps();
    }
```

Pivot Schema : branch_user
```php
        Schema::create('branch_user', function (Blueprint $table) {
            $table->id();
            $table->foreignId('branch_id')->constrained()->onDelete('cascade');
            $table->foreignId('user_id')->constrained()->onDelete('cascade');
            $table->timestamps();
        });
```

in panel provider for branch (Hq is panel for Headquarter's user, backoffice is what i use for branches)
```php
use App\Filament\Hq\Resources\CurriculumResource;
use App\Filament\Hq\Resources\DealResource;
use App\Filament\Hq\Resources\GradeResource;
use App\Filament\Hq\Resources\LevelResource;
use App\Filament\Hq\Resources\ProductResource;
use App\Filament\Hq\Resources\ProductVariantResource;
use App\Filament\Hq\Resources\SchoolResource;
use App\Filament\Hq\Resources\SubjectResource;
use App\Filament\Hq\Resources\UserResource;
use App\Filament\Hq\Resources\VoucherResource;
use App\Filament\Pages\EditProfile;
use App\Http\Middleware\ApplyTenantScopes;

public function panel(Panel $panel): Panel
    {
        return $panel
            ->id('backoffice')
            ->path('backoffice')
            ->login()
            ->profile()
            ->userMenuItems([ 
                'profile' => MenuItem::make()->url(fn (): string => EditProfile::getUrl()),
            ])
            ->colors([
                'primary' => Color::Blue,
            ])
            ->resources([
                ProductResource::class,
                ProductVariantResource::class,
                DealResource::class,
                VoucherResource::class,
                CurriculumResource::class,
                GradeResource::class,
                LevelResource::class,
                SubjectResource::class,
                SchoolResource::class,
                UserResource::class,
            ])
            ->discoverPages(in: app_path('Filament/Backoffice/Pages'), for: 'App\\Filament\\Backoffice\\Pages')
            ->discoverResources(in: app_path('Filament/Resources/Shield'), for: 'App\\Filament\\Resources\\Shield')
            ->pages([
                Pages\Dashboard::class,
                EditProfile::class,
            ])
            ->navigationGroups([
                'Branch Management',
                'Product Management',
                'School Management',
            ])
```
